File: kotlin/matrix/src/main/kotlin/Matrix.kt
class Matrix(matrixAsString: String) {

    companion object {
        private const val ROW_DELIMITER = '\n'
        private const val ELEMENT_DELIMITER = ' '
    }

    private val rows = matrixAsString
            .split(ROW_DELIMITER)
            .map(this::parseStringListToInt)

    fun column(colNr: Int): List<Int> = rows.flatMap(List<Int>::withIndex)
            .filter { it.index == colNr - 1 }
            .map(IndexedValue<Int>::value)

    fun row(rowNr: Int): List<Int> = rows[rowNr - 1]

    private fun parseStringListToInt(row: String) = row.split(ELEMENT_DELIMITER)
            .map(String::trim)
            .filter(String::isNotEmpty)
            .map(Integer::parseInt)
            .toList()
}

File: javascript/matrix/matrix.js
const parseRows = (rowStrings) => rowStrings.map(row => row.split(' ').map(elem => parseInt(elem, 10)))

function generateColumns(maxNumberOfColumns, rows) {
  const columns = []
  for (let i = 0; i < maxNumberOfColumns; ++i) {
    columns.push(rows.map(row => row[i]))
  }

  return columns
}

function parseMatrix(matrixDescription) {
  const rowStrings = matrixDescription.split('\n')
  const rows = parseRows(rowStrings)
  const maxNumberOfColumns = rows[0].length
  const columns = generateColumns(maxNumberOfColumns, rows)

  return {
    rows,
    columns,
  }
}

export class Matrix {
  constructor(matrixDescription) {
    this.matrix = parseMatrix(matrixDescription)
  }

  get rows() {
    return this.matrix.rows
  }

  get columns() {
    return this.matrix.columns
  }
}
